File: lib/repoProfile.ts
/**
 * Phase G — repo profile scanner.
 *
 * Synchronously walks a sibling repo and derives a `RepoProfile`: stack
 * (next, nestjs, prisma, …), domain keywords, high-level features (auth,
 * payments, lms, …), and entrypoint glob patterns. Pure heuristic, no LLM
 * calls — every read is `try/catch` so a half-checked-out / network-broken
 * sibling never crashes the bridge.
 *
 * Two consumers:
 *  - `repoHeuristic.ts` — boosts the keyword scoring with profile signals.
 *  - `coordinator.ts`   — prepends a "## Repo profiles" block to every
 *                         coordinator prompt so the LLM knows the
 *                         contract surface of each candidate repo.
 *
 * TODO: an LLM-assisted upgrade can plug in later via a
 * `summarizeWithLLM(profile)` hook on top of this heuristic baseline.
 */

import {
  existsSync,
  readFileSync,
  readdirSync,
  statSync,
} from "node:fs";
import { basename, join } from "node:path";

export type RouterStyle = "app" | "pages" | "src" | "unknown";
export type PrimaryLang = "ts" | "js" | "py" | "go" | "java" | "unknown";

export interface RepoProfileSignals {
  hasPackageJson: boolean;
  hasReadme: boolean;
  hasClaudeMd: boolean;
  hasNextConfig: boolean;
  hasPrismaSchema: boolean;
  hasTailwindConfig: boolean;
  hasNestCoreDep: boolean;
  hasReactDep: boolean;
  routerStyle: RouterStyle;
  primaryLang: PrimaryLang;
}

export interface RepoProfile {
  name: string;
  path: string;
  summary: string;
  stack: string[];
  keywords: string[];
  features: string[];
  entrypoints: string[];
  fileCounts: Record<string, number>;
  refreshedAt: string;
  signals: RepoProfileSignals;
}

const READ_CAP_BYTES = 4096;
const FILE_WALK_CAP = 5000;
const KEYWORD_CAP = 40;
const TOP_EXTENSIONS_CAP = 5;
const WALK_DEPTH_CAP = 4;

const SKIP_DIRS = new Set([
  "node_modules", ".next", "dist", "build", ".git", "coverage", ".turbo",
  ".cache", ".vercel", ".bridge-state", ".uploads", "out", ".pnpm-store",
]);

const STOPWORDS = new Set([
  "the", "and", "of", "to", "a", "an", "in", "for", "with", "on", "this",
  "that", "is", "are", "be", "by", "as", "or", "at", "from", "it", "its",
  "src", "lib", "app", "public", "dist", "build", "node_modules", "test",
  "tests", "spec", "specs", "pages", "components", "page", "component",
  "config", "configs", "json", "ts", "tsx", "js", "jsx", "md", "mjs",
  "package", "lock", "readme", "license", "docs", "doc",
]);

/**
 * Top-level rules used to derive `features[]`. Order matters only for
 * de-dupe; matching is independent per rule.
 *
 * Each rule scans the union of: deps, top-level dir names, prisma model
 * names, and harvested keywords, looking for any of `match[]` as a
 * substring (lowercased). One hit → feature added once.
 */
const FEATURE_RULES: { feature: string; match: string[] }[] = [
  { feature: "auth",          match: ["auth", "login", "jwt", "oauth", "session"] },
  { feature: "payments",      match: ["payment", "billing", "stripe", "invoice", "subscription"] },
  { feature: "i18n",          match: ["i18n", "locale", "translation", "intl"] },
  { feature: "notifications", match: ["notification", "email", "sms", "mail", "push"] },
  { feature: "messaging",     match: ["chat", "message", "conversation", "thread"] },
  { feature: "lms",           match: ["lms", "course", "lesson", "student", "teacher", "classroom", "quiz", "exam"] },
  { feature: "orchestration", match: ["coordinator", "bridge", "orchestrat", "agent"] },
];

interface ParsedPackageJson {
  deps: Record<string, string>;
  name?: string;
  description?: string;
}

function safeReadText(path: string, capBytes = READ_CAP_BYTES): string | null {
  try {
    if (!existsSync(path)) return null;
    const buf = readFileSync(path);
    return buf.subarray(0, capBytes).toString("utf8");
  } catch {
    return null;
  }
}

function safeReadJson<T>(path: string): T | null {
  try {
    const text = readFileSync(path, "utf8");
    return JSON.parse(text) as T;
  } catch {
    return null;
  }
}

function parsePackageJson(repoPath: string): ParsedPackageJson | null {
  const pkg = safeReadJson<{
    name?: string;
    description?: string;
    dependencies?: Record<string, string>;
    devDependencies?: Record<string, string>;
    peerDependencies?: Record<string, string>;
  }>(join(repoPath, "package.json"));
  if (!pkg) return null;
  const deps: Record<string, string> = {
    ...(pkg.dependencies ?? {}),
    ...(pkg.devDependencies ?? {}),
    ...(pkg.peerDependencies ?? {}),
  };
  return { deps, name: pkg.name, description: pkg.description };
}

function detectRouterStyle(repoPath: string): RouterStyle {
  const hasApp = existsSync(join(repoPath, "app"));
  const hasPages = existsSync(join(repoPath, "pages"));
  const hasSrcApp = existsSync(join(repoPath, "src", "app"));
  const hasSrcPages = existsSync(join(repoPath, "src", "pages"));
  if (hasApp || hasSrcApp) return "app";
  if (hasPages || hasSrcPages) return "pages";
  if (existsSync(join(repoPath, "src"))) return "src";
  return "unknown";
}

function detectPrimaryLang(
  repoPath: string,
  pkg: ParsedPackageJson | null,
): PrimaryLang {
  if (existsSync(join(repoPath, "pyproject.toml")) || existsSync(join(repoPath, "requirements.txt"))) {
    return "py";
  }
  if (existsSync(join(repoPath, "go.mod"))) return "go";
  if (existsSync(join(repoPath, "pom.xml")) || existsSync(join(repoPath, "build.gradle"))) {
    return "java";
  }
  if (pkg) {
    if (pkg.deps["typescript"] || existsSync(join(repoPath, "tsconfig.json"))) return "ts";
    return "js";
  }
  return "unknown";
}

function existsAny(repoPath: string, names: string[]): boolean {
  return names.some((n) => existsSync(join(repoPath, n)));
}

/**
 * Walk the repo from a small set of meaningful roots, counting file
 * extensions. Bounded by `FILE_WALK_CAP` and `WALK_DEPTH_CAP` so a
 * pathological tree can't lock us up.
 */
function countExtensions(repoPath: string): { counts: Record<string, number>; topLevelDirs: string[] } {
  const counts: Record<string, number> = {};
  const topLevelDirs: string[] = [];
  let visited = 0;

  // Collect top-level dir names first (used for keyword harvest).
  try {
    for (const e of readdirSync(repoPath, { withFileTypes: true })) {
      if (!e.isDirectory()) continue;
      if (e.name.startsWith(".")) continue;
      if (SKIP_DIRS.has(e.name)) continue;
      topLevelDirs.push(e.name);
    }
  } catch {
    /* unreadable repo root */
  }

  const walk = (dir: string, depth: number): boolean => {
    if (visited >= FILE_WALK_CAP) return false;
    if (depth > WALK_DEPTH_CAP) return true;
    let entries: import("node:fs").Dirent[];
    try {
      entries = readdirSync(dir, { withFileTypes: true });
    } catch {
      return true;
    }
    for (const e of entries) {
      if (visited >= FILE_WALK_CAP) return false;
      if (e.name.startsWith(".") && e.name !== ".env.example") continue;
      if (SKIP_DIRS.has(e.name)) continue;
      const full = join(dir, e.name);
      if (e.isDirectory()) {
        if (!walk(full, depth + 1)) return false;
      } else if (e.isFile()) {
        visited += 1;
        const dot = e.name.lastIndexOf(".");
        if (dot > 0) {
          const ext = e.name.slice(dot).toLowerCase();
          counts[ext] = (counts[ext] ?? 0) + 1;
        }
      }
    }
    return true;
  };

  walk(repoPath, 0);
  return { counts, topLevelDirs };
}

function topNExtensions(counts: Record<string, number>, n: number): Record<string, number> {
  const out: Record<string, number> = {};
  Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, n)
    .forEach(([ext, count]) => { out[ext] = count; });
  return out;
}

/**
 * Pull the first heading + first non-empty paragraph out of a markdown
 * blob (≤4 KB). Used for the human `summary` field.
 */
function extractMarkdownIntro(md: string): string {
  const lines = md.split(/\r?\n/);
  let heading = "";
  let paragraph = "";
  let inCodeFence = false;
  for (const raw of lines) {
    const line = raw.trim();
    if (line.startsWith("```")) {
      inCodeFence = !inCodeFence;
      continue;
    }
    if (inCodeFence) continue;
    if (!heading && /^#{1,6}\s+/.test(line)) {
      heading = line.replace(/^#{1,6}\s+/, "").trim();
      continue;
    }
    if (heading && line && !line.startsWith("#") && !line.startsWith(">") && !line.startsWith("|")) {
      paragraph = line;
      break;
    }
  }
  if (heading && paragraph) return `${heading} — ${paragraph}`;
  return heading || paragraph || "";
}

/**
 * Read prisma/schema.prisma if present and pluck out `model Foo {…}` /
 * `enum Bar {…}` declarations as keyword candidates.
 */
function readPrismaModels(repoPath: string): string[] {
  const text = safeReadText(join(repoPath, "prisma", "schema.prisma"), 64 * 1024);
  if (!text) return [];
  const out = new Set<string>();
  const re = /^(?:model|enum)\s+([A-Za-z_][A-Za-z0-9_]*)/gm;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    out.add(m[1].toLowerCase());
  }
  return [...out];
}

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/g)
    .filter((t) => t.length >= 3 && !STOPWORDS.has(t) && !/^\d+$/.test(t));
}

function dedupe<T>(arr: T[]): T[] {
  return [...new Set(arr)];
}

function deriveStack(
  pkg: ParsedPackageJson | null,
  repoPath: string,
  primaryLang: PrimaryLang,
): { stack: string[]; signals: Pick<RepoProfileSignals,
    "hasNextConfig" | "hasPrismaSchema" | "hasTailwindConfig"
    | "hasNestCoreDep" | "hasReactDep"> } {
  const stack = new Set<string>();
  const deps = pkg?.deps ?? {};

  const hasNext = !!deps["next"];
  const hasReact = !!deps["react"] || !!deps["react-dom"];
  const hasVue = !!deps["vue"];
  const hasSvelte = !!deps["svelte"];
  const hasNest = !!deps["@nestjs/core"];
  const hasExpress = !!deps["express"];
  const hasTailwind = !!deps["tailwindcss"];
  const hasPrismaDep = !!deps["prisma"] || !!deps["@prisma/client"];
  const hasTypeorm = !!deps["typeorm"];
  const hasAnthropic = !!deps["@anthropic-ai/sdk"];
  const hasPlaywright = !!deps["playwright"] || !!deps["@playwright/test"];

  if (hasNext) stack.add("next");
  // omit react when next is present (redundant)
  if (hasReact && !hasNext) stack.add("react");
  if (hasVue) stack.add("vue");
  if (hasSvelte) stack.add("svelte");
  if (hasNest) stack.add("nestjs");
  if (hasExpress && !hasNext) stack.add("express");
  if (hasTailwind) stack.add("tailwindcss");
  if (hasPrismaDep) stack.add("prisma");
  if (hasTypeorm) stack.add("typeorm");
  if (hasAnthropic) stack.add("anthropic-sdk");
  if (hasPlaywright) stack.add("playwright");

  // File-based confirmation / fallbacks.
  const hasNextConfig = existsAny(repoPath, [
    "next.config.js", "next.config.mjs", "next.config.ts", "next.config.cjs",
  ]);
  if (hasNextConfig) stack.add("next");

  const hasPrismaSchema = existsSync(join(repoPath, "prisma", "schema.prisma"));
  if (hasPrismaSchema) stack.add("prisma");

  const hasTailwindConfig = existsAny(repoPath, [
    "tailwind.config.js", "tailwind.config.mjs", "tailwind.config.ts", "tailwind.config.cjs",
  ]);
  if (hasTailwindConfig) stack.add("tailwindcss");

  if (primaryLang === "py") stack.add("python");
  if (primaryLang === "go") stack.add("go");
  if (primaryLang === "java") stack.add("java");
  if (primaryLang === "ts") stack.add("typescript");

  return {
    stack: [...stack],
    signals: {
      hasNextConfig,
      hasPrismaSchema,
      hasTailwindConfig,
      hasNestCoreDep: hasNest,
      hasReactDep: hasReact,
    },
  };
}

function deriveEntrypoints(
  stack: string[],
  routerStyle: RouterStyle,
  primaryLang: PrimaryLang,
  repoPath: string,
): string[] {
  const out: string[] = [];
  const hasNext = stack.includes("next");
  const hasNest = stack.includes("nestjs");

  if (hasNext) {
    if (routerStyle === "app") {
      out.push("app/api/**/*.ts", "app/**/*.tsx");
    } else if (routerStyle === "pages") {
      out.push("pages/api/**/*.ts", "pages/**/*.tsx");
    } else {
      // unknown / src — best guess
      out.push("app/**/*.tsx", "pages/**/*.tsx");
    }
  }
  if (hasNest) {
    out.push(
      "src/**/*.controller.ts",
      "src/**/*.service.ts",
      "src/**/*.module.ts",
    );
  }
  if (!hasNext && !hasNest) {
    if (primaryLang === "ts" || primaryLang === "js") {
      if (existsSync(join(repoPath, "src"))) out.push("src/**/*.ts");
      if (existsSync(join(repoPath, "lib"))) out.push("lib/**/*.ts");
      if (out.length === 0) out.push("**/*.ts");
    } else if (primaryLang === "py") {
      out.push("**/*.py");
    } else if (primaryLang === "go") {
      out.push("**/*.go");
    } else if (primaryLang === "java") {
      out.push("src/main/java/**/*.java");
    }
  }

  return dedupe(out);
}

function deriveFeatures(haystack: string[]): string[] {
  const blob = haystack.join(" ").toLowerCase();
  const out: string[] = [];
  for (const rule of FEATURE_RULES) {
    if (rule.match.some((needle) => blob.includes(needle))) {
      out.push(rule.feature);
    }
  }
  return dedupe(out);
}

function synthesizeSummary(name: string, stack: string[]): string {
  if (stack.length === 0) return `${name} — repo (no README found, no recognised stack)`;
  const tag = stack.slice(0, 4).join(" + ");
  return `${name} — ${tag} (no README found)`;
}

/**
 * Scan `repoPath` and produce a `RepoProfile`. Always returns; falls
 * back to a synthesized summary + empty stack when there's no
 * package.json / README / etc. Never throws.
 */
export function scanRepo(repoPath: string): RepoProfile {
  const name = basename(repoPath);
  const pkg = parsePackageJson(repoPath);
  const primaryLang = detectPrimaryLang(repoPath, pkg);
  const routerStyle = detectRouterStyle(repoPath);
  const { stack, signals: stackSignals } = deriveStack(pkg, repoPath, primaryLang);

  const claudeMd = safeReadText(join(repoPath, "CLAUDE.md"));
  const readme = safeReadText(join(repoPath, "README.md"));
  const claudeIntro = claudeMd ? extractMarkdownIntro(claudeMd) : "";
  const readmeIntro = readme ? extractMarkdownIntro(readme) : "";
  const summary =
    claudeIntro ||
    readmeIntro ||
    pkg?.description ||
    synthesizeSummary(name, stack);

  const prismaModels = readPrismaModels(repoPath);
  const { counts, topLevelDirs } = countExtensions(repoPath);
  const fileCounts = topNExtensions(counts, TOP_EXTENSIONS_CAP);

  // Keyword harvest: summary + dep names (split on -/_) + prisma models +
  // top-level dir names + pkg name. Lowercase, strip stopwords, dedupe.
  const depTokens = Object.keys(pkg?.deps ?? {}).flatMap((d) =>
    d.replace(/^@/, "").split(/[/_-]+/g),
  );
  const harvest = [
    ...tokenize(summary),
    ...tokenize(depTokens.join(" ")),
    ...prismaModels,
    ...tokenize(topLevelDirs.join(" ")),
    ...(pkg?.name ? tokenize(pkg.name) : []),
  ];
  const keywords = dedupe(
    harvest.filter((w) => w.length >= 3 && !STOPWORDS.has(w)),
  ).slice(0, KEYWORD_CAP);

  const features = deriveFeatures([
    ...keywords,
    ...topLevelDirs,
    ...Object.keys(pkg?.deps ?? {}),
  ]);

  const entrypoints = deriveEntrypoints(stack, routerStyle, primaryLang, repoPath);

  const signals: RepoProfileSignals = {
    hasPackageJson: !!pkg,
    hasReadme: !!readme,
    hasClaudeMd: !!claudeMd,
    routerStyle,
    primaryLang,
    ...stackSignals,
  };

  return {
    name,
    path: repoPath,
    summary,
    stack,
    keywords,
    features,
    entrypoints,
    fileCounts,
    refreshedAt: new Date().toISOString(),
    signals,
  };
}

// Internal helpers exposed for testing only.
export const __test = {
  STOPWORDS,
  FEATURE_RULES,
  tokenize,
  extractMarkdownIntro,
  countExtensions,
  FILE_WALK_CAP,
  KEYWORD_CAP,
};

/**
 * Light placeholder so callers can wire an LLM-driven enrichment later.
 * Returns the profile as-is; replace impl when an SDK is available.
 */
export function summarizeWithLLM(profile: RepoProfile): RepoProfile {
  return profile;
}

/**
 * Helper for callers that want to skip the path/stat dance themselves.
 * `existsCheck = false` short-circuits without touching disk.
 */
export function scanRepoIfExists(
  repoPath: string,
  existsCheck = true,
): RepoProfile | null {
  try {
    if (existsCheck) {
      const st = statSync(repoPath);
      if (!st.isDirectory()) return null;
    }
    return scanRepo(repoPath);
  } catch {
    return null;
  }
}

File: internal/apps/repoprofile.go
package apps

// Heuristic repo profile detector — Go port of libs/repoProfile.ts.
// Pure stdlib, never throws / never returns an error: a half-checked-out
// sibling, a malformed package.json, a missing README — all degrade to
// "empty fields" rather than failing the caller. The bridge calls this
// at app-add time and on operator-driven refresh; failure here would
// strand the whole app registry.
//
// The TS port additionally derived a keyword harvest, file-extension
// counts, prisma model names, and a `signals` block. The S16 follow-up
// scope intentionally omits those — the Go callers (coordinator prompt,
// repo scoring) consume only Stack / Features / Entrypoints / Summary.
// If the wider keyword surface is needed later, port the harvest path
// off of the existing TS module rather than guessing what's used.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// readCapBytes mirrors the TS READ_CAP_BYTES — the README/CLAUDE.md
// summary extractor only ever looks at the first 4 KB so a 50 MB README
// can't blow up memory.
const readCapBytes = 4096

// featureRule maps a feature label to substring needles. Matching is
// independent per rule, one hit -> feature added once. Order is only
// load-bearing for the de-dupe pass downstream.
type featureRule struct {
	feature string
	match   []string
}

var featureRules = []featureRule{
	{"auth", []string{"auth", "login", "jwt", "oauth", "session"}},
	{"payments", []string{"payment", "billing", "stripe", "invoice", "subscription"}},
	{"i18n", []string{"i18n", "locale", "translation", "intl"}},
	{"notifications", []string{"notification", "email", "sms", "mail", "push"}},
	{"messaging", []string{"chat", "message", "conversation", "thread"}},
	{"lms", []string{"lms", "course", "lesson", "student", "teacher", "classroom", "quiz", "exam"}},
	{"orchestration", []string{"coordinator", "bridge", "orchestrat", "agent"}},
}

// RepoProfile is the persisted heuristic snapshot for one repo. The
// shape is the slim subset of libs/repoProfile.ts the Go consumers
// actually read; bridge.json round-trip code does not touch this file.
type RepoProfile struct {
	Name        string   `json:"name"`
	Path        string   `json:"path"`
	Summary     string   `json:"summary"`
	Stack       []string `json:"stack"`
	Features    []string `json:"features"`
	Entrypoints []string `json:"entrypoints"`
	// RefreshedAt is RFC3339Nano so it round-trips with the TS
	// `new Date().toISOString()` shape on operators that already have
	// a repo-profiles.json from the legacy bun build.
	RefreshedAt string `json:"refreshedAt"`
}

// routerStyle is the Next.js routing convention detected in the repo.
// Internal — callers shouldn't care, but it changes which entrypoint
// globs we emit for a Next project.
type routerStyle int

const (
	routerUnknown routerStyle = iota
	routerApp                 // app/ or src/app/
	routerPages               // pages/ or src/pages/
	routerSrc                 // bare src/, no app or pages
)

// primaryLang is the dominant source language. Used to pick fallback
// entrypoint globs and to add a language tag to Stack.
type primaryLang int

const (
	langUnknown primaryLang = iota
	langTS
	langJS
	langPy
	langGo
	langJava
)

// parsedPackageJson is the trimmed package.json shape — only the keys
// the heuristic actually reads.
type parsedPackageJson struct {
	deps        map[string]string
	name        string
	description string
}

// DetectRepoProfile is the pure heuristic entry point. Always returns a
// populated RepoProfile — fields are zero/empty when nothing was
// detected. Never blocks on the network, never opens an LLM, never
// returns an error: callers chain it from cache-refresh paths where a
// per-repo failure must not poison the rest of the batch.
func DetectRepoProfile(name, path string) RepoProfile {
	pkg := parsePackageJson(path)
	lang := detectPrimaryLang(path, pkg)
	router := detectRouterStyle(path)
	stack := deriveStack(pkg, path, lang)
	entrypoints := deriveEntrypoints(stack, router, lang, path)

	// Summary first — features mines its text for orchestration/auth/etc
	// keywords, matching the TS port's behavior of feeding the summary
	// into the feature haystack.
	summary := extractSummary(path, pkg, name, stack)
	features := deriveFeatures(stack, pkg, path, summary)

	return RepoProfile{
		Name:        name,
		Path:        path,
		Summary:     summary,
		Stack:       stack,
		Features:    features,
		Entrypoints: entrypoints,
		RefreshedAt: time.Now().UTC().Format(time.RFC3339Nano),
	}
}

// safeReadText reads up to capBytes from path. Returns "" on any error
// (missing file, perms, …). The TS port distinguished nil vs "" — Go
// callers only ever check `if text != ""`, so collapse to one shape.
func safeReadText(path string, capBytes int) string {
	f, err := os.Open(path)
	if err != nil {
		return ""
	}
	defer f.Close()
	buf := make([]byte, capBytes)
	n, _ := f.Read(buf)
	return string(buf[:n])
}

// parsePackageJson reads + parses package.json, merging dependencies +
// devDependencies + peerDependencies into a single map (the heuristic
// doesn't care which bucket a dep came from).
func parsePackageJson(repoPath string) *parsedPackageJson {
	body, err := os.ReadFile(filepath.Join(repoPath, "package.json"))
	if err != nil {
		return nil
	}
	var raw struct {
		Name             string            `json:"name"`
		Description      string            `json:"description"`
		Dependencies     map[string]string `json:"dependencies"`
		DevDependencies  map[string]string `json:"devDependencies"`
		PeerDependencies map[string]string `json:"peerDependencies"`
	}
	if err := json.Unmarshal(body, &raw); err != nil {
		return nil
	}
	deps := make(map[string]string, len(raw.Dependencies)+len(raw.DevDependencies)+len(raw.PeerDependencies))
	for k, v := range raw.Dependencies {
		deps[k] = v
	}
	for k, v := range raw.DevDependencies {
		deps[k] = v
	}
	for k, v := range raw.PeerDependencies {
		deps[k] = v
	}
	return &parsedPackageJson{deps: deps, name: raw.Name, description: raw.Description}
}

// existsAny is true when any of `names` exists at the root of repoPath.
func existsAny(repoPath string, names []string) bool {
	for _, n := range names {
		if _, err := os.Stat(filepath.Join(repoPath, n)); err == nil {
			return true
		}
	}
	return false
}

func dirExists(p string) bool {
	st, err := os.Stat(p)
	return err == nil && st.IsDir()
}

func detectRouterStyle(repoPath string) routerStyle {
	if dirExists(filepath.Join(repoPath, "app")) || dirExists(filepath.Join(repoPath, "src", "app")) {
		return routerApp
	}
	if dirExists(filepath.Join(repoPath, "pages")) || dirExists(filepath.Join(repoPath, "src", "pages")) {
		return routerPages
	}
	if dirExists(filepath.Join(repoPath, "src")) {
		return routerSrc
	}
	return routerUnknown
}

// detectPrimaryLang prefers explicit signature files (pyproject.toml,
// go.mod, pom.xml) over the package.json fallback so a polyglot repo
// with package.json + go.mod still tags as Go (the more specific
// signal — Node-based tooling lives next to almost every primary lang).
func detectPrimaryLang(repoPath string, pkg *parsedPackageJson) primaryLang {
	if existsAny(repoPath, []string{"pyproject.toml", "requirements.txt"}) {
		return langPy
	}
	if existsAny(repoPath, []string{"go.mod"}) {
		return langGo
	}
	if existsAny(repoPath, []string{"pom.xml", "build.gradle"}) {
		return langJava
	}
	if pkg != nil {
		if _, ok := pkg.deps["typescript"]; ok {
			return langTS
		}
		if existsAny(repoPath, []string{"tsconfig.json"}) {
			return langTS
		}
		return langJS
	}
	return langUnknown
}

// deriveStack inspects package.json deps + filesystem signature files
// and returns a stable de-duped slice of stack tags. A dep-based hit
// can still win even when the file-based confirmation is missing (the
// repo may be brand-new and the operator hasn't run `next init` yet).
func deriveStack(pkg *parsedPackageJson, repoPath string, lang primaryLang) []string {
	seen := make(map[string]bool)
	out := []string{}
	add := func(tag string) {
		if seen[tag] {
			return
		}
		seen[tag] = true
		out = append(out, tag)
	}

	var deps map[string]string
	if pkg != nil {
		deps = pkg.deps
	}
	hasNext := dep(deps, "next")
	hasReact := dep(deps, "react") || dep(deps, "react-dom")
	hasNest := dep(deps, "@nestjs/core")

	if hasNext {
		add("next")
	}
	// react is implied by next — omit to keep the tag set short.
	if hasReact && !hasNext {
		add("react")
	}
	if dep(deps, "vue") {
		add("vue")
	}
	if dep(deps, "svelte") {
		add("svelte")
	}
	if hasNest {
		add("nestjs")
	}
	if dep(deps, "express") && !hasNext {
		add("express")
	}
	if dep(deps, "tailwindcss") {
		add("tailwindcss")
	}
	if dep(deps, "prisma") || dep(deps, "@prisma/client") {
		add("prisma")
	}
	if dep(deps, "typeorm") {
		add("typeorm")
	}
	if dep(deps, "@anthropic-ai/sdk") {
		add("anthropic-sdk")
	}
	if dep(deps, "playwright") || dep(deps, "@playwright/test") {
		add("playwright")
	}

	// File-based confirmation / fallback. Keeps the tag accurate when a
	// dep was hoisted into a workspace package.json the heuristic
	// doesn't see.
	if existsAny(repoPath, []string{
		"next.config.js", "next.config.mjs", "next.config.ts", "next.config.cjs",
	}) {
		add("next")
	}
	if _, err := os.Stat(filepath.Join(repoPath, "prisma", "schema.prisma")); err == nil {
		add("prisma")
	}
	if existsAny(repoPath, []string{
		"tailwind.config.js", "tailwind.config.mjs", "tailwind.config.ts", "tailwind.config.cjs",
	}) {
		add("tailwindcss")
	}

	switch lang {
	case langPy:
		add("python")
	case langGo:
		add("go")
	case langJava:
		add("java")
	case langTS:
		add("typescript")
	}

	return out
}

func dep(deps map[string]string, name string) bool {
	if deps == nil {
		return false
	}
	_, ok := deps[name]
	return ok
}

// deriveFeatures scans the union of (stack tags, dep names, top-level
// dir names) for the FEATURE_RULES needles. The TS port also fed
// keyword + prisma-model harvests; this Go cut leans on dep names +
// dirs because every consumer of features today only cares about the
// rough business-domain bucket.
func deriveFeatures(stack []string, pkg *parsedPackageJson, repoPath, summary string) []string {
	parts := make([]string, 0, len(stack)+8)
	parts = append(parts, stack...)
	if pkg != nil {
		for k := range pkg.deps {
			parts = append(parts, k)
		}
		if pkg.name != "" {
			parts = append(parts, pkg.name)
		}
		if pkg.description != "" {
			parts = append(parts, pkg.description)
		}
	}
	if summary != "" {
		parts = append(parts, summary)
	}
	parts = append(parts, topLevelDirs(repoPath)...)
	blob := strings.ToLower(strings.Join(parts, " "))

	seen := make(map[string]bool)
	out := []string{}
	for _, rule := range featureRules {
		for _, needle := range rule.match {
			if strings.Contains(blob, needle) {
				if !seen[rule.feature] {
					seen[rule.feature] = true
					out = append(out, rule.feature)
				}
				break
			}
		}
	}
	return out
}

// topLevelDirs returns visible (non-dot) top-level directory names.
// Bounded by a single readdir — the heuristic never recurses, so a
// pathological tree depth can't lock us up.
func topLevelDirs(repoPath string) []string {
	entries, err := os.ReadDir(repoPath)
	if err != nil {
		return nil
	}
	out := make([]string, 0, len(entries))
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		if strings.HasPrefix(name, ".") {
			continue
		}
		out = append(out, name)
	}
	return out
}

// deriveEntrypoints picks glob patterns the coordinator can hand to a
// child agent's "look here first" tool list. Next.js + NestJS get
// framework-aware patterns; everything else falls back to a single
// language-keyed glob.
func deriveEntrypoints(stack []string, router routerStyle, lang primaryLang, repoPath string) []string {
	stackHas := func(s string) bool {
		for _, t := range stack {
			if t == s {
				return true
			}
		}
		return false
	}
	out := []string{}
	hasNext := stackHas("next")
	hasNest := stackHas("nestjs")

	if hasNext {
		switch router {
		case routerApp:
			out = append(out, "app/api/**/*.ts", "app/**/*.tsx")
		case routerPages:
			out = append(out, "pages/api/**/*.ts", "pages/**/*.tsx")
		default:
			out = append(out, "app/**/*.tsx", "pages/**/*.tsx")
		}
	}
	if hasNest {
		out = append(out,
			"src/**/*.controller.ts",
			"src/**/*.service.ts",
			"src/**/*.module.ts",
		)
	}
	if !hasNext && !hasNest {
		switch lang {
		case langTS, langJS:
			if dirExists(filepath.Join(repoPath, "src")) {
				out = append(out, "src/**/*.ts")
			}
			if dirExists(filepath.Join(repoPath, "lib")) {
				out = append(out, "lib/**/*.ts")
			}
			if len(out) == 0 {
				out = append(out, "**/*.ts")
			}
		case langPy:
			out = append(out, "**/*.py")
		case langGo:
			out = append(out, "**/*.go")
		case langJava:
			out = append(out, "src/main/java/**/*.java")
		}
	}
	return dedupeStrings(out)
}

func dedupeStrings(in []string) []string {
	seen := make(map[string]bool, len(in))
	out := make([]string, 0, len(in))
	for _, s := range in {
		if seen[s] {
			continue
		}
		seen[s] = true
		out = append(out, s)
	}
	return out
}

// extractSummary picks the human-readable one-liner. Preference order
// matches the TS port: CLAUDE.md > README.md > package.json description >
// synthesized fallback. The fallback exists so a wholly-undocumented
// repo still surfaces *something* in the coordinator prompt.
func extractSummary(repoPath string, pkg *parsedPackageJson, name string, stack []string) string {
	if intro := extractMarkdownIntro(safeReadText(filepath.Join(repoPath, "CLAUDE.md"), readCapBytes)); intro != "" {
		return intro
	}
	if intro := extractMarkdownIntro(safeReadText(filepath.Join(repoPath, "README.md"), readCapBytes)); intro != "" {
		return intro
	}
	if pkg != nil && pkg.description != "" {
		return pkg.description
	}
	if len(stack) == 0 {
		return name + " — repo (no README found, no recognised stack)"
	}
	tag := strings.Join(firstN(stack, 4), " + ")
	return name + " — " + tag + " (no README found)"
}

func firstN(in []string, n int) []string {
	if len(in) <= n {
		return in
	}
	return in[:n]
}

// extractMarkdownIntro pulls "first heading — first paragraph" out of a
// markdown blob. Skips fenced code blocks, blockquotes, table rows.
// Empty input -> empty output (caller's preference chain falls through).
func extractMarkdownIntro(md string) string {
	if md == "" {
		return ""
	}
	lines := strings.Split(strings.ReplaceAll(md, "\r\n", "\n"), "\n")
	heading := ""
	paragraph := ""
	inFence := false
	for _, raw := range lines {
		line := strings.TrimSpace(raw)
		if strings.HasPrefix(line, "```") {
			inFence = !inFence
			continue
		}
		if inFence {
			continue
		}
		if heading == "" && isMarkdownHeading(line) {
			heading = strings.TrimSpace(strings.TrimLeft(line, "# "))
			continue
		}
		if heading != "" && line != "" &&
			!strings.HasPrefix(line, "#") &&
			!strings.HasPrefix(line, ">") &&
			!strings.HasPrefix(line, "|") {
			paragraph = line
			break
		}
	}
	if heading != "" && paragraph != "" {
		return heading + " — " + paragraph
	}
	if heading != "" {
		return heading
	}
	return paragraph
}

// isMarkdownHeading is true for `#` … `######` followed by space + text.
// Avoids the regexp dep just for one shape.
func isMarkdownHeading(line string) bool {
	if !strings.HasPrefix(line, "#") {
		return false
	}
	i := 0
	for i < len(line) && line[i] == '#' && i < 6 {
		i++
	}
	return i >= 1 && i <= 6 && i < len(line) && line[i] == ' '
}
